pyam/core.py

```diff
@@ -317,28 +326,26 @@ def categorize(self, name, value, criteria,
             if arg:
                 run_control().update({kind: {name: {value: arg}}})
 
-        if criteria == 'uncategorized':
-            self.meta[name].fillna(value, inplace=True)
-            msg = "{} of {} scenarios are uncategorized."
-            logger().info(msg.format(np.sum(self.meta[name] == value),
-                                     len(self.meta)))
-            return  # EXIT FUNCTION
-
         # find all data that matches categorization
         rows = _apply_criteria(self.data, criteria,
                                in_range=True, return_test='all')
         idx = _meta_idx(rows)
 
-        # update metadata dataframe
-        if name not in self.meta:
-            self.meta[name] = np.nan
         if len(idx) == 0:
             logger().info("No scenarios satisfy the criteria")
-        else:
-            self.meta.loc[idx, name] = value
-            msg = '{} scenario{} categorized as `{}: {}`'
-            logger().info(msg.format(len(idx), '' if len(idx) == 1 else 's',
-                                     name, value))
+            return  # EXIT FUNCTION
+
+        # update metadata dataframe
+        self._add_meta_column(name, value)
+        self.meta.loc[idx, name] = value
+        msg = '{} scenario{} categorized as `{}: {}`'
+        logger().info(msg.format(len(idx), '' if len(idx) == 1 else 's',
+                                 name, value))
+
+    def _add_meta_column(self, name, value):
```

Comment:
Ok, if I understand this correctly, I would call this function `_new_meta_column()` and have it with a kwarg like `fill_nan=True`. Then you could call it with `_new_meta_column(name, fill_nan=not isstr(value))`.

You can obviously play with the fill_nan or fill_uncategorized as you prefer, but it wasn't clear how the value was being used in the calling function.

Reply:
agree on the renaming, but for the arg, the whole point was to move the checking for the type to the auxiliary function, not into the args when calling the function